### PR TITLE
added SupervisorSpec CRD

### PR DESCRIPTION
**Pull request by applike-ss**
Fixes https://github.com/druid-io/druid-operator/issues/313

### Description

We're currently managing our supervisor specs as json files, however that does not seem to be ideal - especially in a Infrastructure as Code world where having it applied should only be one command away.

To make our Supervisor specs available for IaC tools, i have crafted this change-set to allow creating SupervisorSpec CRs and apply them to the cluster in the defined state.
This is my first try to hack some support of any kind into an operator in kubernetes, so there may be some things done plain wrong and wrong terminology used in code.

This Change was done in a hackathon under time pressure, so please bear with me as the code will not be perfect.
Let me know what you think anyways.

I'll rebase this at some point, for now i just wanted to make sure to create this PR.

<hr>

This PR has:
- [ ] been tested on a real K8S cluster to ensure creation of a brand new Druid cluster works.
- [x] been tested on a real K8S cluster to ensure creation/modification/removal of a SupervisorSpec works.
- [x] been tested for backward compatibility on a real K*S cluster by applying the changes introduced here on an existing Druid cluster. If there are any backward incompatible changes then they have been noted in the PR description.
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added documentation for new or modified features or behaviors.

<hr>

##### Key changed/added files in this PR
 * CRD added (SupervisorSpec)
 * .tool-versions file added for asdf support
 * added zookeeper/mysql to `helm-install-druid-operator` make job for easier one-shot installs (may be removed before merge)
 * added basic cluster.yaml which was used for testing (may be removed before merge)
 * SupervisorSpec management code

### CODE REVIEW

**Comment by AdheipSingh** [comment]
@applike-ss IMHO you can add your PR into the workflow design [here](https://github.com/datainfrahq/druid-operator/pull/53) . 
Also can you point to the custom resource example for the supervisor spec.

For batch, supervisor, sql we plan to have a single controller which can support all of them.

**Comment by applike-ss** [comment]
I've added my example here: https://github.com/datainfrahq/druid-operator/pull/61/commits/342a330fe89436e3a730d57cfa314c41d86eb12c

**Comment by applike-ss** [comment]
If i get your CR and code right, then your change is for one-time tasks while mine is for supervisor specs, so long-running stateful tasks. Is this correct?

What do you mean with adding my PR into the workflow design "there"? Do you mean i should adjust my PR description to match the structure of yours?

**Comment by AdheipSingh** [comment]
@applike-ss thank you. This PR is really helpful, and looking at the CR we are very much aligned. 

Lets sync on [druid-operator slack](https://kubernetes.slack.com/?redir=%2Farchives%2FC04F4M6HT2L) for faster iteration. 
You can join #druid-operator channel on kubernetes slack.

File: .tool-versions
Comment (itamar-marom) [inline .tool-versions:1]: What is this file for?

File: Makefile [hunk 1]
```diff
@@ -189,8 +189,17 @@ helm-install-druid-operator: ## helm upgrade/install
 	helm upgrade --install \
 	--namespace ${NAMESPACE_DRUID_OPERATOR} \
 	--create-namespace \
+	zookeeper bitnami/zookeeper
```
Comment (itamar-marom) [inline Makefile:192]: ZooKeeper should be deployed on another namespace - not in the druid operator namespace

File: Makefile [hunk 2]
```diff
@@ -189,8 +189,17 @@ helm-install-druid-operator: ## helm upgrade/install
...
+	--namespace ${NAMESPACE_DRUID_OPERATOR} \
+	--create-namespace \
+	--set auth.username=druid \
+	--set auth.password=druid \
+	--set auth.database=druid \
+	mysql bitnami/mysql
+	helm upgrade --install \
+	--namespace ${NAMESPACE_DRUID_OPERATOR} \
 	${NAMESPACE_DRUID_OPERATOR} chart/ \
-	--set image.repository=${IMG_KIND} \
```
Comment (itamar-marom) [inline Makefile:193]: Why was this removed?

File: apis/druid/v1alpha1/supervisorspec_types.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+ */
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
```
Comment (itamar-marom) [inline apis/druid/v1alpha1/supervisorspec_types.go:11]: please remove this and the below unnecessary comments

File: chart/values.yaml [hunk 4]
```diff
@@ -12,7 +12,7 @@ replicaCount: 1
 
 image:
   repository: datainfrahq/druid-operator
-  pullPolicy: IfNotPresent
+  pullPolicy: Always
```
Comment (itamar-marom) [inline chart/values.yaml:15]: Why always?

File: deploy/operator.yaml [hunk 5]
```diff
@@ -16,7 +16,7 @@ spec:
       containers:
         - name: druid-operator
           # Replace this with the built image name
-          image: datainfrahq/druid-operator:latest
+          image: 164105964448.dkr.ecr.eu-central-1.amazonaws.com/struss-test-druid-operator:latest
```
Comment (itamar-marom) [inline deploy/operator.yaml:19]: change it back? :)

File: cluster.yaml [hunk 6]
```diff
@@ -0,0 +1,453 @@
+apiVersion: druid.apache.org/v1alpha1
```
Comment (itamar-marom) [inline cluster.yaml:1]: Should be removed

File: supervisor-spec.yaml [hunk 7]
```diff
@@ -0,0 +1,183 @@
+apiVersion: druid.apache.org/v1alpha1
```
Comment (itamar-marom) [inline supervisor-spec.yaml:1]: Should be removed

File: controllers/druid/supervisorspec_controller.go [hunk 8]
```diff
@@ -0,0 +1,608 @@
...
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the SupervisorSpec object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.6.4/pkg/reconcile
+func (r *SupervisorSpecReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithName(string(uuid.NewUUID()))
+	_ = log.WithValues("supervisorspec", req.NamespacedName)
```
Comment (itamar-marom) [inline controllers/druid/supervisorspec_controller.go:90]: can be removed?

File: controllers/druid/supervisorspec_controller.go [hunk 9]
```diff
@@ -0,0 +1,608 @@
...
+	log.Info(fmt.Sprintf("reconciling SupervisorSpec (%s)", req.NamespacedName.String()))
+
+	action := ActionCreate
+	supervisorCr := &druidv1alpha1.SupervisorSpec{}
+	err := r.Client.Get(ctx, req.NamespacedName, supervisorCr)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			log.Error(err, "failed to get SupervisorSpec from k8s api")
+		}
+		action = ActionDelete
```
Comment (itamar-marom) [inline controllers/druid/supervisorspec_controller.go:101]: If I get it right, this is how you handle Kubernetes object deletion? You shouldn't handle it that way. You should check if Kubernetes added to this object a `DeletionTimestamp`.

Also, what happens if the Druid cluster is deleted?

File: controllers/druid/supervisorspec_controller.go [hunk 10]
```diff
@@ -0,0 +1,608 @@
...
+	default:
+		log.Error(fmt.Errorf("unexpected action: %s", action), "Error occurred")
+	}
+
+	log.Info(fmt.Sprintf("reconciled SupervisorSpec (%s)", req.NamespacedName.String()))
+
+	return ctrl.Result{}, nil
+}
+
+func (r *SupervisorSpecReconciler) objectKeyFromStringSlice(input []string, fallbackNamespace string) client.ObjectKey {
```
Comment (itamar-marom) [inline controllers/druid/supervisorspec_controller.go:145]: General function, shouldn't be part of this struct

**Comment by itamar-marom** [comment]
@applike-ss Are you up to finishing this? This PR can be useful.

File: .tool-versions
Comment (applike-ss) [reply .tool-versions:1]: This is for the ASDF tool version manager. That is a shell based application that installs shims for the tools you wanna use and based on which directory you are in uses the correct tool version. Say you are involved in >10 Projects and each is using a different version of kubebuilder as per the projects policy. Now you don't have to worry about that anymore and can just do a `asdf install` in the projects directory (after you added the plugin the first time with `asdf plugin add kubebuilder`) and you are ready to rock.

File: chart/values.yaml (hunk 4 above)
Comment (applike-ss) [reply chart/values.yaml:15]: i think, i did this for my development purposes where i killed the container to "reload" the operator with new code

File: deploy/operator.yaml (hunk 5 above)
Comment (applike-ss) [reply deploy/operator.yaml:19]: dang.... i thought i found all "internal references". Definitely changing this back

File: controllers/druid/supervisorspec_controller.go (hunk 8 above)
Comment (applike-ss) [reply controllers/druid/supervisorspec_controller.go:90]: why? It (log) is setting the value on itself inside WithValues via `l.setSink(l.sink.WithValues(keysAndValues...))`.

**Comment by applike-ss** [comment]
> @applike-ss Are you up to finishing this? This PR can be useful.

There's no doubt i would like to finish it up, but i will need to schedule this with my lead. I'm doing this within business hours.

File: controllers/druid/supervisorspec_controller.go (hunk 8 above)
Comment (itamar-marom) [reply controllers/druid/supervisorspec_controller.go:90]: log.WithValues() return a new logger

**Comment by applike-ss** [comment]
I am currently really packed with work and will not be able to finish this in the upcoming month(s).

**Comment by applike-ss** [comment]
FYI: i am proceeding with this PR. Currently setting up my test env and trying to get my PR back into a working state as some things have changed meanwhile

File: Makefile (hunk 2 above)
Comment (applike-ss) [reply Makefile:193]: i've adjusted it back for now. Switched to developing the app locally with telepresence v2, so i have easy access to services and the k8s api.

File: apis/druid/v1alpha1/supervisorspec_types.go (hunk 3 above)
Comment (applike-ss) [reply apis/druid/v1alpha1/supervisorspec_types.go:11]: This does exist for the druid spec types as well, should i remove those too?

File: chart/values.yaml (hunk 4 above)
Comment (applike-ss) [reply chart/values.yaml:15]: reverted

File: controllers/druid/supervisorspec_controller.go (hunk 8 above)
Comment (applike-ss) [reply controllers/druid/supervisorspec_controller.go:90]: You're right, i did see a pointer where there wasn't any.

**Comment by AdheipSingh** [comment]
@applike-ss i would like to discuss the design with you. Please schedule a call on my [calendly](https://calendly.com/adheip-singh/30-min-meeting) anytime. I would want to have some core changes in the spec definition.

**Comment by applike-ss** [comment]
IDK if it makes sense to sync currently.
i have brought it back in a working state now and removed the status "InSync", as it was hard to keep working and might easily break in the near future.
My employer did only want to grant me 1-2 days of dev time which i used 1 day already to create a somewhat working local test env.
So my fear is that we sync now and waste time because i can't implement everything in time.
I will try to finish this in the future, but that might be rather months than weeks.

**Comment by AdheipSingh** [comment]
we have ingestion controller now. feel free to contribute to it. 
thanks
